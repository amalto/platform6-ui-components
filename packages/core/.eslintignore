dist/
loader/

src/components.d.ts