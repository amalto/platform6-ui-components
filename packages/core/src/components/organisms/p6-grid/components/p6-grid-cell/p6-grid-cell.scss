$table-cell-padding: 0.5em 0.75em !default;

:host {
  display: flex;
  position: relative;
  box-sizing: border-box;
  padding: $table-cell-padding;
  vertical-align: middle;
  font-size: $size-normal;
  font-family: $family-primary;
}

:host(:not(:last-child)) {
  width: 100px;
}
