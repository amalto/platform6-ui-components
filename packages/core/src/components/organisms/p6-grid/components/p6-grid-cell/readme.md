# p6-grid-cell

<!-- Auto Generated Below -->

## Properties

| Property              | Attribute  | Description                                           | Type                                                   | Default           |
| --------------------- | ---------- | ----------------------------------------------------- | ------------------------------------------------------ | ----------------- |
| `align`               | `align`    | Cell alignment                                        | `Alignment.center \| Alignment.end \| Alignment.start` | `Alignment.start` |
| `color`               | `color`    | Cell text color                                       | `string \| undefined`                                  | `undefined`       |
| `column` _(required)_ | --         | Cell column                                           | `Column<Record<string, unknown>>`                      | `undefined`       |
| `editable`            | `editable` | Set to true if this cell is editable, otherwise false | `boolean`                                              | `false`           |
| `row` _(required)_    | --         | Cell line                                             | `Row<Record<string, unknown>>`                         | `undefined`       |
| `width`               | `width`    | Cell width                                            | `number \| string`                                     | `DEFAULT_WIDTH`   |

## Dependencies

### Used by

- [p6-grid](../..)

### Depends on

- [p6-grid-textarea-cell-editor](../p6-grid-textarea-cell-editor)

### Graph

```mermaid
graph TD;
  p6-grid-cell --> p6-grid-textarea-cell-editor
  p6-grid --> p6-grid-cell
  style p6-grid-cell fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
