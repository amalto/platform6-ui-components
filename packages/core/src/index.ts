export { Components, JSX } from './components';
export { P6ButtonType } from './components/atoms/p6-button/p6-button';
export { P6CalendarType } from './components/atoms/p6-calendar/p6-calendar';
export { P6CheckboxValue } from './components/atoms/p6-checkbox/p6-checkbox';
export { P6InputValue } from './components/atoms/p6-input/p6-input';
export { P6InputType } from './components/atoms/p6-input/types';
export { Target } from './components/atoms/p6-link/p6-link';
export { P6RadioValue } from './components/atoms/p6-radio/p6-radio';
export { P6SelectControl, P6SelectValue } from './components/atoms/p6-select-native/p6-select-native';
export { P6SwitchValue } from './components/atoms/p6-switch/p6-switch';
export { P6TextareaValue } from './components/atoms/p6-textarea/p6-textarea';
export { P6LanguageValue } from './components/molecules/p6-language/p6-language';
export { P6TranslationControl, P6TranslationValue } from './components/molecules/p6-translation/p6-translation';
export {
  AlignColumnDetail,
  CellValueChangedDetail,
  Column,
  ColumnDefinition,
  ColumnId,
  DataItem,
  EditingCellDetail,
  FilterRowsDetail,
  MoveColumnDetail,
  ResetDefinitionsDetail,
  ResizeColumnDetail,
  Row,
  RowId,
  ShowColumnDetail,
  ShowOptionsDetail,
  SortColumnDetail,
} from './components/organisms/p6-grid/core/entities';
export { ServiceStatus } from './components/templates/p6-service-panel/service-status.enum';
export { P6Control } from './shared/form/control';
export { InvalidEventDetail, isCustomEvent, isInvalidEvent, isValidEvent, ValidEventDetail } from './shared/form/event';
export { Alignment, Mode, modes, Position, positions, Size, sizes, SortOrder } from './shared/types';
export { EnumArrayEntry, enumArrayToObject, enumToArray, EnumType } from './utils/enum';
export { LanguageCode } from './utils/language';
