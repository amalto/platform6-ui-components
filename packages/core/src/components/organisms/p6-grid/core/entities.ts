import { EventEmitter, JSX } from '@stencil/core';
import { Alignment, Direction, SortOrder } from '../../../../shared/types';

export type ColumnId = string;

export type DataItem = Record<string, unknown>;

export interface ColumnDefinition<T extends DataItem> {
  id?: ColumnId;
  label: string;
  field?: string | keyof T;
  editable?: boolean;
  filtreable?: boolean;
  sortable?: boolean;
  hidden?: boolean;
  width?: number;
  align?: Alignment;
  color?: string;
  sortOrder?: SortOrder;
  disableHeaderMenu?: boolean;
  getValue?: (data: T, context: ColumnDefinition<T>) => string;
  setValue?: (value: string, data: T, context: ColumnDefinition<T>) => T;
  cellEditor?: (data: T, context: ColumnDefinition<T>) => JSX.Element & CellEditor<T>;
  cellRenderer?: (data: T, rowId: RowId, context: ColumnDefinition<T>) => string | HTMLElement;
}

export interface Column<T extends DataItem> extends ColumnDefinition<T> {
  id: ColumnId;
  editable: boolean;
  hidden: boolean;
  width: number;
  align: Alignment;
  sortOrder: SortOrder;
  disableHeaderMenu: boolean;
  getValue: (data: T, context: ColumnDefinition<T>) => string;
  setValue: (value: string, data: T, context: ColumnDefinition<T>) => T;
}

export interface CellEditor<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  value: string;

  p6GridCellValueChanged: EventEmitter<CellValueChangedDetail<T>>;
  p6GridEditingCell: EventEmitter<EditingCellDetail<T>>;
}
export type RowId = string;

export interface Row<T extends DataItem> {
  id: RowId;
  data: T;
}

interface ColumnEventDetail<T extends DataItem> {
  column: Column<T>;
}

export interface MoveColumnDetail<T extends DataItem> extends ColumnEventDetail<T> {
  direction: Direction;
}
export type ResizeColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type AlignColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type SortColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type ShowColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export interface CellValueChangedDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  oldValue: string;
}

export enum EditingCellStatus {
  Start,
  Stop,
  Cancel,
}

export interface EditingCellDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  status: EditingCellStatus;
}

export interface ResetDefinitionsDetail {
  reset: boolean;
}

export interface ShowOptionsDetail {
  visible: boolean;
}

export interface FilterRowsDetail {
  value: string;
}
