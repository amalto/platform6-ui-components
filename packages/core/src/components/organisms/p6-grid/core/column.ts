import { Alignment, Direction, SortOrder } from '../../../../shared/types';
import { throwError } from '../../../../utils/error';
import { Column, ColumnDefinition, ColumnId, DataItem } from './entities';

export const MIN_WIDTH = 75;
export const DEFAULT_WIDTH = MIN_WIDTH;
export const INC_WIDTH = 10;

export function defaultGetValue<T extends DataItem>(data: T, context: ColumnDefinition<T>): string {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'getValue' function for ${JSON.stringify(context)}`);
  }

  return `${data[context.field as keyof T]}`;
}

export function defaultSetValue<T extends DataItem>(value: string, data: T, context: ColumnDefinition<T>): T {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'setValue' function for ${JSON.stringify(context)}`);
  }

  return {
    ...data,
    [context.field as keyof T]: value,
  };
}

let columnIds = 0;
export function fromDefinition<T extends DataItem>(column: ColumnDefinition<T>): Column<T> {
  // eslint-disable-next-line no-plusplus
  const id = column.id || column.field || columnIds++;

  return {
    ...column,
    id: `col-${id}`,
    editable: column.editable || false,
    hidden: column.hidden || false,
    width: column.width || 100,
    align: column.align || Alignment.center,
    sortOrder: column.sortOrder || SortOrder.none,
    disableHeaderMenu: column.disableHeaderMenu || false,
    getValue: column.getValue || defaultGetValue,
    setValue: column.setValue || defaultSetValue,
  };
}

export function isHidden<T extends DataItem>(column: Column<T>): boolean {
  return !!column.hidden;
}

export function replaceColumn<T extends DataItem>(updatedColumn: Column<T>, allColumns: Column<T>[]): Column<T>[] {
  return allColumns.map(col => (col.id === updatedColumn.id ? updatedColumn : col));
}

export function move<T extends DataItem>(id: ColumnId, direction: Direction, columns: Column<T>[]): Column<T>[] {
  const from = columns.findIndex(col => col.id === id);
  const to = from + (direction === Direction.left ? -1 : 1);

  if (from < 0 || to < 0 || to >= columns.length) {
    return columns;
  }

  const result = columns.slice();
  const item = result.splice(from, 1);

  return result.slice(0, to).concat(item).concat(result.slice(to, columns.length));
}
