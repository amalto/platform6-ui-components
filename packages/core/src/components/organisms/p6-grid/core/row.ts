import { SortOrder } from '../../../../shared/types';
import { Column, DataItem, Row, RowId } from './entities';

let rowIds = 0;
export function fromData<T extends DataItem>(data: T): Row<T> {
  rowIds += 1;

  return {
    id: `row-${rowIds}`,
    data,
  };
}

export function compareRow<T extends DataItem>(column: Column<T> | undefined): (a: Row<T>, b: Row<T>) => number {
  const sortOrder = column?.sortOrder || SortOrder.none;
  const factor = sortOrder === SortOrder.desc ? -1 : 1;

  return (a: Row<T>, b: Row<T>): number => {
    if (column === undefined || column.sortable === false || sortOrder === SortOrder.none) {
      return 0;
    }

    const valueA = column.getValue(a.data, column);
    const valueB = column.getValue(b.data, column);

    return valueA.localeCompare(valueB) * factor;
  };
}

export function filterBy<T extends DataItem>(searchValue: string, columns: Column<T>[]): (row: Row<T>) => boolean {
  return row => columns.filter(col => col.filtreable !== false).some(col => col.getValue(row.data, col).includes(searchValue));
}

export function rangeSelectRow<T extends DataItem>(currentRowIdSelect: RowId, selectedRows: Set<RowId>, displayedRow: Row<T>[]): Set<RowId> {
  const currentSelectedIndex = displayedRow.findIndex(row => row.id === currentRowIdSelect);

  if (currentSelectedIndex === -1) {
    return selectedRows;
  }

  const previousRowIdSelected = Array.from(selectedRows)
    .reverse()
    .find(rowId => displayedRow.findIndex(row => row.id === rowId) > -1);

  if (previousRowIdSelected === undefined) {
    return new Set(selectedRows.add(currentRowIdSelect));
  }

  const previousSelectedIndex = displayedRow.findIndex(row => row.id === previousRowIdSelected);

  let [from, to] = [previousSelectedIndex, currentSelectedIndex];
  let excludedFromSlice = currentRowIdSelect;
  if (currentSelectedIndex < previousSelectedIndex) {
    [from, to] = [to, from];
    excludedFromSlice = previousRowIdSelected;
  }

  return displayedRow
    .slice(from, to)
    .reduce((selected, cur) => {
      return cur.id === undefined ? selected : selected.add(cur.id);
    }, new Set(selectedRows))
    .add(excludedFromSlice);
}

export function replaceRow<T extends DataItem>(updatedRow: Row<T>, allRows: Row<T>[]): Row<T>[] {
  return allRows.map(row => (row.id === updatedRow.id ? updatedRow : row));
}
