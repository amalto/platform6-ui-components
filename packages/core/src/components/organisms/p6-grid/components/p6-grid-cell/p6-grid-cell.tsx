import { Component, ComponentInterface, Element, h, Host, Listen, Prop, State } from '@stencil/core';
import { Alignment } from '../../../../../shared/types';
import { toWidth } from '../../../../../utils/css';
import { DEFAULT_WIDTH } from '../../core/column';
import { Column, DataItem, EditingCellDetail, EditingCellStatus, Row } from '../../core/entities';

@Component({
  tag: 'p6-grid-cell',
  styleUrl: 'p6-grid-cell.scss',
  shadow: true,
})
export class P6GridCell implements ComponentInterface {
  @Element() host!: HTMLP6GridCellElement;

  /**
   * Cell column
   */
  @Prop() column!: Column<DataItem>;

  /**
   * Cell line
   */
  @Prop() row!: Row<DataItem>;

  /**
   * Cell alignment
   */
  @Prop() align: Alignment = Alignment.start;

  /**
   * Cell text color
   */
  @Prop() color: string | undefined;

  /**
   * Cell width
   */
  @Prop() width: number | string = DEFAULT_WIDTH;

  /**
   * Set to true if this cell is editable, otherwise false
   */
  @Prop() editable = false;

  /**
   * Set to true if this cell is being edited
   */
  // eslint-disable-next-line @stencil/strict-mutable
  @Prop({ mutable: true }) editing = false;

  @State() isEditing = false;

  @Listen('p6GridEditingCell')
  onP6GridEditingCell(event: CustomEvent<EditingCellDetail<DataItem>>): void {
    if (event.detail.status !== EditingCellStatus.Start) {
      this.isEditing = false;
    }
  }

  componentWillLoad(): void {
    this.host.addEventListener('dblclick', () => {
      if (this.editable) {
        this.isEditing = true;
      }
    });
  }

  render(): JSX.Element {
    const { align, color, width } = this;
    const styles = {
      color,
      justifyContent: Alignment[align],
      width: toWidth(width),
    };
    const classes = {
      'is-editing': this.isEditing,
    };

    return (
      <Host class={classes} style={styles}>
        {this.isEditing ? this.getCellEditor(this.row, this.column) : <slot />}
      </Host>
    );
  }

  // eslint-disable-next-line class-methods-use-this
  private getCellEditor(row: Row<DataItem>, column: Column<DataItem>) {
    return column.cellEditor !== undefined ? (
      column.cellEditor(row.data, column)
    ) : (
      <p6-grid-textarea-cell-editor row={row} value={column.getValue(row.data, column)} column={column} />
    );
  }
}
