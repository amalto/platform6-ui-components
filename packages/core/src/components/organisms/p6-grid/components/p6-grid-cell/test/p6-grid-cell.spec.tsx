import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { fromDefinition } from '../../../core/column';
import { Column, DataItem, Row } from '../../../core/entities';
import { P6GridCell } from '../p6-grid-cell';

describe('p6-grid-cell', () => {
  const column: Column<DataItem> = fromDefinition({
    id: 'colId',
    label: 'Col Label',
    editable: true,
    width: 150,
    getValue: () => '',
  });

  const row: Row<DataItem> = {
    id: '1',
    data: {},
  };

  it('should render an empty cell', async () => {
    const page = await newSpecPage({
      components: [P6GridCell],
      template: () => {
        return <p6-grid-cell column={column} row={row} />;
      },
    });
    expect(page.root).toMatchSnapshot();
  });
});
