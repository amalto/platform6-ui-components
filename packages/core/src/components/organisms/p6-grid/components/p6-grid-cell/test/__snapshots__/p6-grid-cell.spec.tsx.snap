// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-grid-cell should render an empty cell 1`] = `
<p6-grid-cell style="justify-content: start; width: 75px;">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
</p6-grid-cell>
`;
